libs/ddns/src/update.rs

```diff
@@ -0,0 +1,307 @@
+    let dhcid_record = Record::from_rdata(
+        name.clone(),
+        0,
+        RData::Unknown {
+            code: 49,
+            rdata: NULL::with(duid.rdata(&name)?),
+        },
+    );
+    message.add_pre_requisite(dhcid_record);
+
+    let a_record = Record::from_rdata(name, ttl, RData::A(leased));
+    message.add_update(a_record);
+
+    Ok(message)
+}
+
+pub fn delete(
+    zone_origin: Name,
+    name: Name,
+    duid: DhcId,
+    leased: Ipv4Addr,
+    ttl: u32,
+    use_edns: bool,
+) -> Result<trust_dns_client::op::Message, NameError> {
+    use trust_dns_client::{
+        op::UpdateMessage,
+        rr::{rdata::NULL, RData, Record, RecordType},
+    };
+
+    let rev_ip = Name::from_str(&reverse_ip(leased)).unwrap();
+    let mut message = update_msg(zone_origin, use_edns);
+
+    // delete
+    let owner = Record::with(rev_ip.clone(), RecordType::ANY, 0);
+    let dhcid = Record::with(rev_ip.clone(), RecordType::ANY, 0);
+    message.add_update(owner);
+    message.add_update(dhcid);
+    // add
+    let ptr_record = Record::from_rdata(rev_ip.clone(), ttl, RData::PTR(name.clone()));
+    let dhcid_record = Record::from_rdata(
+        rev_ip,
+        ttl,
+        RData::Unknown {
+            code: 49,
+            rdata: NULL::with(duid.rdata(&name)?),
+        },
+    );
+    message.add_update(ptr_record);
+    message.add_update(dhcid_record);
+
+    Ok(message)
+}
+
+fn update_msg(zone_origin: Name, use_edns: bool) -> trust_dns_client::op::Message {
+    use trust_dns_client::{
+        op::{Edns, Message, MessageType, OpCode, Query, UpdateMessage},
+        rr::{DNSClass, RecordType},
+    };
+    const MAX_PAYLOAD_LEN: u16 = 1232;
+
+    let mut zone = Query::new();
+    zone.set_name(zone_origin)
+        .set_query_class(DNSClass::IN)
+        .set_query_type(RecordType::SOA);
+
+    let mut message = Message::new();
+    message
+        .set_id(rand::random())
+        .set_message_type(MessageType::Query)
+        .set_op_code(OpCode::Update)
+        .set_recursion_desired(false);
+
+    message.add_zone(zone);
+
+    if use_edns {
+        let edns = message.extensions_mut().get_or_insert_with(Edns::new);
+        edns.set_max_payload(MAX_PAYLOAD_LEN);
+        edns.set_version(0);
+    }
+
+    message
+}
+
+pub fn reverse_ip<I: Into<IpAddr>>(ip: I) -> String {
+    let ip = ip.into();
+    match ip {
+        IpAddr::V4(ip) => {
+            let [a, b, c, d] = ip.octets();
+            format!("{}.{}.{}.{}.in-addr.arpa.", d, c, b, a)
+        }
+        IpAddr::V6(ip) => {
+            let mut s = ip
+                .octets()
+                .iter()
+                .rev()
+                .map(|o| {
+                    // convert u8 into reverse nibbles
+                    // 1 byte fa -> "a.f"
+                    let a = char::from_digit(((*o >> 4) & 0xF) as u32, 16).unwrap();
+                    let b = char::from_digit((*o & 0xF) as u32, 16).unwrap();
+                    format!("{b}.{a}")
```

Comment:
```suggestion
                    format!("{o:02x}")
```

Reply:
this suggestion doesn't reverse the nibbles, unless I am missing something? for ex.
`188 -> c.b`
with format!("{o:02x}")
`188 -> bc`

Reply:
Oops, you are right